### DISCUSSION TITLE
Extending Wearables schema and defining other entities (emotes, third party, etc)

### DISCUSSION THREAD

**Discussion by guidota** [decentraland/adr#45] [General]
## Description

Adding more and more entities into the catalysts sounds good and the way to go, so we need to make it easy and generic enough to get flexibility.

Emotes are very similar to wearables at first sight, but they can become very complex (sequencing, for instance). Initial implementation would be fine to extend Wearables schema to include some fields like `emoteData` or more representations.

Some questions arise among others: 
- Do we need a new entity type?
- Will Emotes be included in other entities like Scenes, Portable Experiences, Wearables? 
- How can we distinguish Emotes from Wearables? 

We briefly discussed about designing a framework where we can compose entities and do schema versioning (retro compatibility?), some thoughts would be ideal to move fast (fail fast?) in order to implement Emotes for Fashion Week, but also taking into account this will be preceding other features.

@pentreathm @menduz @AjimenezDCL @olavra 

Other entities in mind that can be affected by this discussion will be Third Party items. 

## Questions

 Every time we want to add new entities we need to think in these questions:
 - Who will create (deploy) them?: To start, I'd use the same mechanism as the Smart Wearables, which is to upload a zip to the builder. That would require tho, a change in builder to accept the new metadata for the Emote. https://github.com/orgs/decentraland/teams/dapps
- What will the URN/identifier be?: Will use a collections-based URN
- What will the max content size be?: To start it would inherit the size of the wearables. Later, if needed we can add extra space if the wearable is composed with an emote
- How do we test the emotes? Using the same mechanism provided by builder: [ADR-53](https://github.com/decentraland/adr/blob/e14d0898e5de5507ea2277f6f7a5f227e0d20635/docs/ADR-53-test-collections-in-the-explorer.md)

## Open Questions

- What does it have in common with other entities? (Just to take into account when designing entities)
- How many entities will be active in the catalyst?:

 
 (extending invitation to @decentraland/dapps @decentraland/platform)

**Comment by menduz** [comment]
I think we can start clarifying some questions, assuming we do the Proof of concept on top of the wearables to leverage all the business flows like curators, collections, marketplace, lambdas, entities in catalysts and subgraphs among other.

- Who will create (deploy) them?: To start, I'd use the same mechanism as the Smart Wearables, which is to upload a zip to the builder. That would require tho, a change in builder to accept the new metadata for the Emote. @decentraland/dapps 
- What will the URN/identifier be?: Will use a collections-based URN
- What will the max content size be?: To start it would inherit the size of the wearables. Later, if needed we can add extra space if the wearable is composed with an emote
- How many entities will be active?: This will be defined by both the backpack and the renderer, since it is purely visual [ADR-57](https://github.com/decentraland/adr/blob/aefe59553e9d403df729a018e8da4bea039228e2/docs/ADR-57-avatar-assembling-for-visualization.md)
- How do we test the emotes? Using the same mechanism provided by builder: [ADR-53](https://github.com/decentraland/adr/blob/e14d0898e5de5507ea2277f6f7a5f227e0d20635/docs/ADR-53-test-collections-in-the-explorer.md)
- What does it have in common with other entities?: Can you be more specific? Which use cases are you foreseeing?

**Reply by guidota** [reply]
> What does it have in common with other entities?: Can you be more specific? Which use cases are you foreseeing?

Is a question to think while designing new stuff, nothing to respond, just to avoid having multiple definitions for same things.

**Reply by AjimenezDCL** [reply]
>How many entities will be active?

Could you be a bit more specific on this question? "Active" as equipped at a time? 

If we just mean active in the backpack for visualization it will depend or the renderer itself, the target platform and such.

**Reply by menduz** [reply]
@guidota I'd like to propose that you update the initial first posting with the answered questions and with the open questions to prevent having them sparse over the comments

**Reply by guidota** [reply]
Active => the deployment is part of the last snapshot and is actually referenced by any pointer (users, parcels). It cannot be deleted by the Garbage Collector (if activated).

For instance, `Profiles`: some deployments are just history since the users can update the profile and previous profiles can be deleted. 
 So, 1 user <=> 1 active profile (for the time being, this can change in the future)

`User stores` for example can be deployed by every user but are optional, so we can have M stores and N users (where M < N).

Wearables are not limited but we know there are a few thousands.
Scenes are relative to parcels.

Aprox. amount of entities per type:
<img width="196" alt="Screen Shot 2022-02-03 at 10 08 14" src="https://user-images.githubusercontent.com/7695773/152348791-3212d4ad-af0e-443d-8e81-34c1e3faf552.png">
 
 Understanding how many active deployments are there, we can deduct the disk min or max space usage for the entity type.

**Reply by AjimenezDCL** [reply]
Sorry for the delay, I didnt notice the reply due to being collapsed in the "thread".

I'm afraid I cannot answer how many active entities will we have. Maybe it would be helpful to specify a bit more the question to avoid mistake it for how many equipped emotes can somebody have.

**Comment by menduz** [comment]
@AjimenezDCL friendly reminder for follow up

**Comment by menduz** [comment]
@guidota would you like to take the chance to do some schema composability/versioning initiative? or may we move forward with an open schema?

**Reply by guidota** [reply]
Let's start with the open Wearable schema and I will come up with some alternatives, Emotes feature should not be blocked.

**Reply by AjimenezDCL** [reply]
On the _renderer_ side reusing the wearables for the emotes doesn't affect us (at least negatively). In fact we can take advantage of the architecture we have for the wearables catalog.

To kick-off emotes I dont think we need this discussion to be done but having at least a stub of the wearable schema will prevent us from backtracking too much.

**Reply by AjimenezDCL** [reply]
Just for the record a candidate schema for the Emotes M1, is still WIP but the composition would work along these lines.

```json
{
  "id": "headexplode",
  "name": "Head Explode",
  "description": "A fitting and engaging description",
  "rarity": "epic",
  "tags": null,
  "emoteV0Data": {
    "loop": false
  },
  "thumbnail": "QmMyThumbnailHash",
  "representations": [
    {
      "bodyShapes": [
        "urn:decentraland:off-chain:base-avatars:BaseFemale"
      ],
      "mainFile": "f_head_explode.glb",
      "contents": [
        {
          "file": "f_head_explode.glb",
          "hash": "QmMyHashValue"
        }
      ]
    }
  ]
}
```

**Comment by LautaroPetaccio** [comment]
For the TP wearables, we have a slightly different schema than for the current one (Wearable).
We have, for the metadata, a new property, called merkelTreeProof, containing the `index` in the tree where the item is and the `hashes` that can be used, along a computed hash (computed at the time of being inserted) to prove that is can be inserted. All of these property `must not be missing` because they're required to compute its validity.
The TP wearables don't have the properties `collectionAddress` nor `rarity` as those are concepts of Wearables that belong to standard collections.

For the TPWearables and the standard Wearables, we compute the hash of the entity, that is, we hash the entity's contents and its metadata to later do the corresponding checks taking into consideration which is the type of wearable. For the TPWearables it will be computed to check that it belongs to a merkle tree and for the Wearables it will be used to check that they're the content hash of an item in the blockchain. The metadata of the entities hashed MUST contain the same properties, that is, we must be sure that what we hash in the dApps side is the same that it's being hashed in the catalyst side.

**Reply by guidota** [reply]
In this case setting `additionalProperties` as `true` is not enough but I think creating a new entity type (or schema) for Third Party Wearables is too much. Maybe adding some defaults to the `collectionAddress` and `rarity` ?

What do you think @menduz ?

(ref: https://github.com/decentraland/common-schemas/pull/54)

**Reply by guidota** [reply]
There is another problem with the `merkleTreeProof`. It can't be part of the metadata since it uses the hash of the entity, and to hash the entity we need the complete metadata. One idea was to move the proof to a file and treat it as content, but we are in the same situation as the content has the files with respective hashes.

In conclusion, the proof can't be part of the entity but it should be part of the deployment. 

cc @nachomazzara

**Reply by nachomazzara** [reply]
Hmmm, The `collectionAddress` and `rarity` are really important for a wearable representation BUT they should not be for third party items. In order to avoid confusion and make it less error prune is it possible to check the urn and decide to use a Wearable or a TPWearable schema? Maybe we can use the same entity but use different schemas to validate based on the urn. If that is not possible I would then suggest going with another entity. It will make the code more clear even we need to add boilerplate

**Reply by menduz** [reply]
@LautaroPetaccio I want to check if I understood correctly. Something in the logic doesn't seem right if this is true:

```typescript
// A
const entity = { ...tpwEntity }
entity.merkelTreeProof = calculate(entity)

// B
const entity = { ...tpwEntity, merkelTreeProof: calculate(tpwEntity.contents) }
```

Does the algorithm works like the case A or the case B?

---

@nachomazzara `collectionAddress` and `rarity` are not really important for the catalyst nor explorer. We can dispose them, I'd recommend to perform the validations in the schema as you suggest, using the URN to check whether or no the fields are available.

Adding a new entity type is not easy and definitely not convenient. The schemas should be as composable as possible and we may check the capabilities (using URN, schema versions or tags) to get the information we need out of the entities.

Following that line, I recommend you to add the merkle validations to a sub-key of the entity object to include a single key like this

```ts
// before
type Wearable = {
  wearableData: any
  merkleTreeProof: MerkleProof
  merkleTreeIndex: number
}

// after
type MerkleProofInformation = {
  merkleTreeProof: MerkleProof
  merkleTreeIndex: number
}
type Wearable = {
  merkle?: MerkleProofInformation
}
```

**Reply by LautaroPetaccio** [reply]
My comment is outdated. The `merkleTreeProof` shouldn't be part of the metadata, as we're using the metadata to produce the "content hash" or the hash of the wearable entity. I thought initially that we could put the `merkleTreeProof` in the metadata to later remove it before computing the wearable's hash, but that won't be the case. As I was told, the idea is to send the `merkleTreeProof` as something else, but that was being discussed (@guidota might know if there were any discussions about this). 

The Catalyst will receive when deploying the entity, a `merkleTreeProof`, containing the `index` and the `hashes` and it will calculate the hash of the wearable entity as it is doing right now with standard collection. All of these properties are needed to verify if the wearable is part of the merkle tree which has its merkle tree root published in the contract.

**Reply by menduz** [reply]
Wait a sec, there is nothing else than the entity itself to embed any information, other methods would require _huge hairy refactors_, as the authChain could be. Why was sending the `merkleTreeProof` inside the entity discarded? as per ADR32 we can only calculate the hashes of the content files to perform the validations. Was that taken into account? @guidota @LautaroPetaccio

**Reply by guidota** [reply]
That's the problem, we have no other information than the entity and authChain in a deployment. As I understand, we need the entire entity to hash and use it on the `merkleTreeProof` (case A), correct me if I'm wrong, but it won't work.

In case B, I think it would work but not sure if `tpwEntity.contents` is enough for the `merkleTreeProof`(I guess metadata is needed too).

Maybe we can store the `merkleTreeProof` in the entity and use partial entity information to calculate the hash (contents + metadata + ?)

**Reply by menduz** [reply]
I bet on that approach, case B. We can make it as elegant and scalable as possible. It would be a better way to do [ADR32 section 2](https://github.com/decentraland/adr/blob/main/docs/ADR-32-wearable-committee-reverts.md#2-use-the-hashmetadata--content).

For maximum flexibility, it can also be expanded to 

```ts
type MerkleRootCapable = {
  merkleInformation?: {
    proof: any
    index: number
    includedFields: string[]
    hashedContent: string
  }
}
type Wearable = MerkleRootCapable & { ... }

// calculates the merkle root and embeds it into the entity
function calculateMerkleRoot(mut entity: MerkleRootCapable, includedFields: string[])

// validates the hashed content and merkle proof
// 1. using entity.merkleInformation.includedFields, generate object X
// 2. serialize object X (deterministic) in Serialized(X)
// 3. hash(Serialized(X)), compare against entity.merkleInformation.hashedContent for CRC
// 4. validate merkle proof of the hash
function validateMerkleRoot(mut entity: MerkleRootCapable, extra)
``` 


I'd prevent modifying the authChain at all costs due to exploding complexity

**Reply by nachomazzara** [reply]
+1 to the idea of adding the `merkleTreeProof` inside the entity but not as part of the metadata. so the ADR32 will continue working as it is. At least that is what I had in mind.

**Reply by menduz** [reply]
@nachomazzara I think we are confusing the terms here, the entity file _is_ the metadata file. And then the entityId is the CIDv1 of that metadata file
